[refman] Clarify parsing and notation interpretation phases (#19453)
coq/coq: doc/sphinx/user-extensions/syntax-extensions.rst

Syntax extensions and notation scopes
=====================================

Syntax extensions let you customize how some nonterminal symbols
in Coq's grammar are entered and displayed.  Syntax extensions include:

- :ref:`Notations <Notations>`, which customize :n:`@term`\s.  They
  let you use the familiar syntax :n:`1+2*3` instead of
  :n:`(Nat.add 1 (Nat.mul 2 3))`.  (Coq accepts both forms; the :flag:`Printing Notations`
  flag controls which is displayed.)

  Notation definitions let you specify:
  
  - *Precedence* so that :n:`1+2*3` is interpreted

>     in the usual way as :n:`1+(2*3)` rather than as `(1+2)*3`.

Reviewer: ```suggestion
    in the usual way as :n:`1+(2*3)` rather than as :n:`(1+2)*3`.
```

>     in the usual way as :n:`1+(2*3)` rather than as `(1+2)*3`.

Reviewer: ```suggestion
    in the usual way as :n:`1+(2*3)` rather than as :n:`(1+2)*3`.
```

Syntax extensions and notation scopes
=====================================

While, it's in theory possible to write down every Gallina term
explicitly, in the way they are understood by the kernel, for instance
:n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`, this would be very

> tedious and error-prone and take us away from our usual mathematical

Reviewer: Unlike academic papers, a chapter introduction in user documentation should first answer the questions "Is this material worth reading now?" and "Why do I care?"  The introduction should be brief and only identify the most important capabilities/concepts (somewhat akin to an abstract but not identical).

Introductions should be written to minimize what you need to know to understand them.  We want it to be accessible to people who are Coq beginners and who aren't already very familiar with Coq.  We shouldn't assume that users have an academic background in proof theory.  IMO the material should be accessible to someone with a bachelor's degree in math or computer science.

These lines introduce terms that are not essential to using notations: "Gallina", "kernel".  It would be better not to use them here.  The text is also wordy, e.g. "it's in theory possible to write down every Gallina term
explicitly, in the way they are understood by the kernel" is a lotta words.  It could be a lot more succinct, making it easier to read.

Chapter and section introductions are often critical for readers/learners to navigate the documentation.  Writing good introductions takes time and care.  And the topic sentence needs to be simple and well-crafted.

practice. To circumvent this, Coq offers notation mechanisms that
reflect the eponym mathematical practice and allows for instance to
write :n:`1 + 2 * 2` instead. Those mechanisms range from simple
:ref:`Abbreviations` to full fledged :ref:`Notations` with user
defined :ref:`syntaxes <ReservingNotations>`. Multiple interpretations
can be given to the same notation in different contexts thanks to the
:ref:`scope <Scopes>` mechanism. For instance :n:`(1 + 2 * 2)%N` could
denote the above natural number expression while :n:`(1 + 2 * 2)%Z`

> could denote a similar expression with integers.

Reviewer: You mention several mechanisms (abbreviations, notations, syntaxes, scopes) but don't (briefly) say what they do/how they are useful or (for most) give tiny examples, which is precisely what an introduction should do.  Hard to make much sense of "multiple _interpretations_ can be given to the same notation in different _contexts_" without knowing what the italicized terms mean.  The doc overuses "context" with multiple meanings; it isn't very informative.

In the introduction text I pushed yesterday, I was careful to minimize the introduction of new terms and to give concrete examples for each new item.

parses and prints objects, i.e. the translations between the concrete
and internal representations of terms and commands.

> When inputing a term to Coq, it goes through several successive steps.

Reviewer: Better to make a more specific claim here than to punt at the end with "elaborating and type checking come later but are out of scope for the current section."  As a reader, I would be pretty disappointed.

Author: Tried to improve that, let me know if you have a better solution.

Author: Sorry, I had to do a rebase to introduce the "moving things around" commits that you, rightfully, required and then forgot to make temporary change commits. In any case, clicking on the "force pushed" link from Github is a good way to get a diff in those cases.

* lexing: the input is split in a linear stream of tokens, for
  instance the string ``"1 + 2 * 2"`` will be split into the sequence
  of five tokens `'1' '+' '2' '*' '2'`

* parsing: the previous stream is given a tree-like structure, for

>   instance here

Reviewer: The parser doesn't actually construct a parse tree data structure.  Rather, as each grammar production is recognized (corresponding to non-leaf nodes in the parse tree), the parser calls the corresponding action function.  The tree is conceptually useful for thinking about, but better to be clear that it's just a concept, not the implementation.

Author: That's not actually true. Keep in mind, that this chapter is not about the parser in general, camlp5, mlg files or what not. It's only about the notation mechanism. There is indeed some kind of action function inside the notation mechanism but its only role is to build the tree. So I don't think there is any point mentioning this internal detail to the reader here.

Reviewer: The parser does not build a parse tree data structure.  (I'm quite familiar with the parsing code.)  Would you point me to the code that you believe "gives a tree-like structure"?  Are you thinking of some code specific to notations?

I suggest tweaking the wording so it's clear that the tree is a (useful) concept and not suggest there is a tree data structure.  When the doc gives a simplified explanation of something, it should still be precise and accurate as far as it goes.

> So I don't think there is any point mentioning this internal detail to the reader here.

Agreed, however descriptions generally need to take into account much more detail than will appear in the doc.  I certainly wouldn't mention action routines here, but their use influences what is said.

Author: > The parser does not build a parse tree data structure. (I'm quite familiar with the parsing code.)

Bear in mind that this chapter of the refman is not about the generic parser but about the specifics of the Notation mechanism. And while, indeed, generic parsers could, in theory, do anything else than building an AST, they usually do and in our particular case it definitely does (and does nothing else).

> Would you point me to the code that you believe "gives a tree-like structure"? Are you thinking of some code specific to notations?

https://github.com/coq/coq/blob/a7f51315db5d70888af3b96a579eb799a2b45ca9/interp/constrexpr.mli#L170

(and the rule building it is basically the `make_act` function in egramcoq.ml)

> I suggest tweaking the wording so it's clear that the tree is a (useful) concept and not suggest there is a tree data structure. When the doc gives a simplified explanation of something, it should still be precise and accurate as far as it goes.

No, as shown above, the tree definitely exists, it's not a simplified inacurate explanation by any mean. And I stand by the fact that it being built by some internal camlp5 rule (the `make_act` function mentionned above) is a technical detail of absolutely no interest for the reader.

  .. code-block:: text
     :name: after-parsing

             *
            / \
           1   +
              / \
             2   2

* notation interpretation: each syntactic element gets translated to a
  term, for instance :n:`1` can be interpreted as the natural number
  :n:`S O` then :n:`2` is interpreted as :n:`S (S O)`, then :n:`2 * 2`
  as :n:`Nat.mul (S (S O)) (S (S O))` and finally our whole term as

>   :n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`.

Reviewer: This only applies to translating single token notations (and maybe only numbers).

Multitoken notations are implemented by adding additional rules to the parser grammar.  They are not implemented as operations on the parse tree.  For example,

```
(* in Notations.v )
Reserved Notation "x <= y" (at level 70, no associativity).
Reserved Notation "x < y" (at level 70, no associativity).
Reserved Notation "x >= y" (at level 70, no associativity).
Reserved Notation "x > y" (at level 70, no associativity).
```

generates the following productions in the grammar:

```
Print Grammar.

(*
Entry term is
  :
| "70" RIGHTA
  [ SELF; ">"; NEXT
  | SELF; ">="; NEXT
  | SELF; "<"; NEXT
  | SELF; "<="; NEXT
*)
```

There is additional post-parsing logic that distinguishes based on the notation scope (e.g. '+' in nat vs Z vs Reals)

Author: > This only applies to translating single token notations (and maybe only numbers).

Could you elaborate? In what sense are "x + y" or "x * y" single token notations? Maybe the example would be clearer with variables rather than constants? or mixing both?

> Multitoken notations are implemented by adding additional rules to the parser grammar. They are not implemented as operations on the parse tree. For example,
[...]
> generates the following productions in the grammar:
[...]

Sure but I'm not sure the detailled syntax of camlp5 is of any interest in this introductory paragraph. That syntax only appears through the `Print Grammar` which is a pretty advanced command (and even more so now that we have `Print Notation`). That syntax is documented there but I'm not sure it is worth introducing it anywhere else.

> There is additional post-parsing logic that distinguishes based on the notation scope (e.g. '+' in nat vs Z vs Reals)

Indeed, that paragrah doesn't even mention scopes. Maybe it should?

parses and prints objects, i.e. the translations between the concrete
and internal representations of terms and commands.

> When inputing a term to Coq, it goes through several successive steps.

Author: This makes a lot of sense. However I don't know what's the best way of implementing it here. In particular since the explanation is needed to understand the following such paragraph (even before this PR it was already mentioning that abbreviation "does not modify the parser"). What should we do
* add yet another such introductory paragraph, about what notations can be used for, before the explanation
* just permute the explanation and the following "sectioning" paragraph, even if it means getting the definitions after their first use
* something else?

> When inputing a term to Coq, it goes through several successive steps.

Author: Thanks, I did try that.

> * :gdef:`lexing`: the input is split in a linear stream of tokens, for

Reviewer: ```suggestion
* :gdef:`lexing`: the input is split into a series of tokens, for
```

If one is going to define a term "lexing" that appears in the glossary index, it should refer to the [Lexical Conventions](https://coq.gitlabpages.inria.fr/-/coq/-/jobs/4641769/artifacts/_build/default/doc/refman-html/language/core/basic.html#lexical-conventions) section rather than to this extremely terse definition.  The LC section has 3-4 pages of detailed information.

Author: I agree. However, `Lexical Conventions` only defines the tokens, not lexing itself. What should we do? I guess we should add a definition of lexing (it does not need to be much more than this sentence) in lexical conventions and refer to it here.

  instance the string ``"1 + 2 * 2"`` will be split into the sequence
  of five tokens `'1' '+' '2' '*' '2'`

* :gdef:`parsing`: the previous stream is given a tree-like structure, for

>   instance here

Reviewer: Similarly, if "parsing" is listed in the glossary index, this terse and incomplete reference is the wrong place to define it.

Author: Agree but not sure what would be the best place to put that definition.

  .. code-block:: text
     :name: after-parsing

             *
            / \
           1   +
              / \
             2   2

> * notation :gdef:`interpretation`: each syntactic element gets translated to a

Reviewer: Again, this is not the place to define "interpretation".  It deserves its own section, which shouldn't be in the Notations chapter.

Author: Well, `interpretation` is a notation specific notion so it should definitely be defined in this chapter. Maybe the definition should come with the `Notation` command? and this would just be a forward reference?

  term, for instance :n:`1` can be interpreted as the natural number
  :n:`S O` then :n:`2` is interpreted as :n:`S (S O)`, then :n:`2 * 2`
  as :n:`Nat.mul (S (S O)) (S (S O))` and finally our whole term as
  :n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`.

Each step is performed independently and there isn't any kind of

> backtracking from any step to a previous one. Then, the result goes

Reviewer: First sentence is unclear and needs to be reworked.  The negative statement "there isn't any kind of backtracking" isn't nearly as helpful as a positive statement of what Coq does.  The steps are distinct but inherently interleaved: Coq lexes a bit, parses a bit, interprets a bit and then repeats the cycle.

Author: Well, not at all, the three phases are absolutely successive and none has any influence on the previous ones. That's what I try to make clear here. Obviously I failed. Any suggestion welcome (I agree negative statements are not ideal).

Reviewer 2: I think perhaps the subtlety here is that while the steps are independent, they are not independent passes over the entire document, right? The interpretation phase of `Reserved Notation` can modify the lexer for subsequent statements.  Perhaps starting with "On each document sentence, each step is performed sequentially and independently." or similar?

through the remaining of the proof assistant, that is the elaboration

> and type checking phases discussed everywhere else in this manual. No

Reviewer: Alas, the doc doesn't have a good description of elaboration or type checking.  Or maybe fair to say "any description".

Author: Well, "not any description" is a bit unfair, since we have at least the typing rules. Anyway, that's unrelated to the current PR.

Reviewer: The typing rules as presented are not understandable to those who are not already very familiar with proof logic.  There's no explanation of the notation they're written in, really no explanation of how to compute the type of an expression or do typechecking and no examples.

Not something to fix in this PR, but "type checking phases discussed elsewhere" is misleading here given the current state of the doc.

types are involved at any point during the above notation handling
phases. And reciprocally, no notation remains during the later type
checking phases. This chapter introduces commands enabling to
customize those notation steps.

> We first introduce :ref:`abbreviations <Abbreviations>`, a kind of
> macro which does not modify the parser.

Reviewer 3: Is this better?
```suggestion
We first introduce :ref:`abbreviations <Abbreviations>`.
These are sometimes called “macros” in other programming language.
They merely associate an identifier (possibly with syntactic variables) with a replacement text.
```

> macro which does not modify the parser.

Reviewer: Defining "abbreviations" simply as "macros" is not informative.  Should describe how they're used.

"does not modify the parser" - not helpful for the user, also this is the first mention of modifying the parser

Author: I agree it's not ideal but the fact that this has nothing to do with the parser is quite crucial. Not sure what's the best way to state it. Suggestions welcome.

Reviewer 4: It's more about (not) modifying the grammar which is parsed than modifying the engine which parses it (ie the parser) imo

More elaborate custom notations can also be defined by modifying the
lexer and parser thanks to the :cmd:`Reserved Notation` command, then
giving interpretations to the new syntax via the :cmd:`Notation`
command. The :cmd:`Infix` command provides a shortcut to the latter
for the specific case of infix notations. It is sometimes expected
that the same syntax has different meanings in different contexts; to
achieve this form of overloading, Coq offers a notion of
:ref:`notation scopes <Scopes>`.

Finally, the main command to provide custom notations for tactics is
:cmd:`Tactic Notation`.

.. coqtop:: none

   Set Printing Depth 50.

.. _Abbreviations:

Abbreviations

> -------------

Author: >  We should try to order sections based on their importance to the user.

I do agree with that and,

> [abbreviations are] considerably less significant for users.

while I don't have any precise figure, abbreviations are in practice used quite a lot, maybe as much if not more than full fledged notations. At least, they are by no means a rare subfeature.

> Notation command should [...] mention [...] Abbreviations as a special case.

Well, in practice it's much more an orthogonal feature than a special case. Abbreviation don't change the syntax, don't deal with scopes, apart from the `Notation` keyword and the fact that they are both handled in the same interpretation phase, the two features don't have much in common.

> Nor does it make the explanation of regular notations any simpler

Sure but the previous state did make explanation of abbreviations much more complicated.

> -------------

Reviewer: Notations are absolutely pervasive, abbreviations are not.  (People use '+', etc. pervasively.)  Coq would be much harder to use without them.  Abbreviations are useful but not as critical.

Just like notations, abbreviations are way of substituting parameters into a replacement expression.  The doc should describe their differences.

If you think it's truly an orthogonal feature, then why put it first?  It uses no syntax extensions or scopes, so maybe it doesn't even belong in the "Syntax extensions and notation scopes" chapter?  (hyperbole)

   and it also can be referred to by a qualified name.

   Abbreviations are syntactic in the sense that they are bound to
   expressions which are not typed at the time of the definition of the
   abbreviation but at the time they are used. Especially, abbreviations
   can be bound to terms with holes (i.e. with “``_``”). For example:

   .. coqtop:: none reset

      Set Strict Implicit.
      Set Printing Depth 50.

   .. coqtop:: in

      Definition explicit_id (A:Set) (a:A) := a.

   .. coqtop:: in

      Notation id := (explicit_id _).

   .. coqtop:: all

      Check (id 0).

   Abbreviations disappear when a section is closed. No typing of the
   denoted expression is performed at definition time. Type checking is
   done only at the time of use of the abbreviation.

   Like for notations, if the right-hand side of an abbreviation is a
   partially applied constant, the abbreviation inherits the implicit
   arguments and notation scopes of the constant. As an
   exception, if the right-hand side is just of the form :n:`@@qualid`,
   this conventionally stops the inheritance of implicit arguments.

>    Like for notations, it is possible to bind binders in
>    abbreviations. Here is an example:

Reviewer 3: ```suggestion
   Like for notations, it is possible to bind (syntactic) variables in
   abbreviations. Here is an example:
```

   .. coqtop:: in reset

      Definition force2 q (P:nat*nat -> Prop) :=
        (forall n', n' >= fst q -> forall p', p' >= snd q -> P q).

      Notation F p P := (force2 p (fun p => P)).
      Check exists x y, F (x,y) (x >= 1 /\ y >= 2).

> 

Reviewer 3: Warnings and errors are missing (I don't know if it was already the case before).
For instance, what if we define a macro whose name is already taken (say, by a definition)?
Which other non-locality attributes are supported?

Author: Indeed, this part of the manual is missing a whole lot of errors. Adding them probably deserve its own PR.

.. _ReservingNotations:

Reserving notations
-------------------

> .. cmd:: Reserved Notation @string {? ( {+, @syntax_modifier } ) }

Reviewer 3: Your comment is valid. I get @proux01's point though: by doing so we separate the "lexing+parsing" phase from the interpretation. Maybe we can keep it that way and add simple examples in the introduction.

> .. cmd:: Reserved Notation @string {? ( {+, @syntax_modifier } ) }

Author: Not that there is actually no such thing as `defining notations without reserving them`, it's just that the `Notation` command currently silently reserves notations for you if it wasn't already done.

Indeed, there are very good rationales for making the parsing / interpretation phase separation more explicit (case in point, your comments in this PR indicate that you didn't fully got it with the previous state of the manual). Not just the fact that the phases are in practice separated but also the fact that they are of very different nature in terms of compositionallity and risks. The reserving notations / parsing things is one of the least modular aspects of Coq. An unfortunate grammar extension can easily wreak havoc in code importing it and seriously compromise compositionality between diferent libraries (although the import categories enabling to import a module without its grammar extensions is a very nice workaround to make progress in that direction). In contrary notation interpretation is much easier to override thanks to the scope mechanism and doesn't present nearly as much risk in terms of compositionality.

I agree however that maybe we should add more examples, potentially using the `Notation` command, in the reserving notation section, so as to be able to fully demonstrate parsing and printing results.

> .. cmd:: Reserved Notation @string {? ( {+, @syntax_modifier } ) }

Reviewer: > Not that there is actually no such thing as `defining notations without reserving them`, it's just that the `Notation` command currently silently reserves notations for you if it wasn't already done.

No.  Think in terms of what the user has to do, not in terms of internal operations.  They can define notations with a single command.  If they want to overload the syntax, they must do a reserve first.

> An unfortunate grammar extension can easily wreak havoc in code importing it

This applies equally to notations defined solely with Notation and those defined with Reserved Notation.

> case in point, your comments in this PR indicate that you didn't fully got it with the previous state of the manual

This comment is not helpful.  If you think I'm mistaken, you should discuss it with me and explain the rationale rather then writing me off.  I have no idea what you think I'm mistaken about.  I hope our comments here are considered a discussion aimed at reaching a consensus (and sometimes mutual education) rather than an argument to win or lose.

   Modifies the lexer (by adding keywords) and the parser.

> Notations must be in double quotes, except when the
> abbreviation has the form of an ordinary applicative expression;
> see :ref:`Abbreviations`. The notation consists of *tokens* separated by

Reviewer 3: I think it's better to keep with the PR's tendency to make abbreviations and notations different beasts.
```suggestion
Notations must be in double quotes, unlike :ref:`Abbreviations`.
The notation consists of *tokens* separated by
```

introduced with a lot of care.

.. cmd:: Reserved Infix @string {? ( {+, @syntax_modifier } ) }

   This command declares an infix :term:`parsing` rule.

When a format is attached to a reserved notation (with the `format`
:token:`syntax_modifier`), it is used by default for printing all
subsequent interpretations of the corresponding
notation. Individual interpretations can override the format.

.. warn:: Notations "a b" defined at level x and "a c" defined at level y have incompatible prefixes. One of them will likely not work.
   :name: notation-incompatible-prefix

   The two notations have a common prefix but different levels.
   The levels of one of the notations should be adjusted to match
   the other. See :ref:`factorization <NotationFactorization>` for
   details.

.. _PrecedencesAndAssociativity:

Precedences and associativity
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Mixing different symbolic notations in the same text may cause serious
:term:`parsing` ambiguity. To deal with the ambiguity of notations, Coq uses
precedence levels ranging from 0 to 100 (plus one extra level numbered
200) and associativity rules.

Consider for example the new notation

.. coqtop:: in

>    Reserved Notation "A \/ B".

Reviewer: What's the rationale for changing all the examples to use `Reserved Notation`?  I don't see the point.  Elsewhere in the chapter the doc can explain when it's best practice or required to use reservations.  It could also say in one place that the examples often use just `Notation` for simplicity.

I think it's helpful for readers to see what the notation is transformed into (ie `(or A B)`).

Author: Agree, c.f. above comment.

symbol starts with a double quote, it must be surrounded with single
quotes to prevent confusion with the beginning of a string symbol.

> Here are examples from the initial state of Coq.

Reviewer 3: (I can't comment parts that did not change but this seems important).
The example above does not work for me (maybe I misunderstood):
```Coq
Reserved Notation "A ""I'm an """"infix"""" symbol"" B" (at level 199, right associativity).
```
gives me:
```
String '"I'm an ""infix"" symbol"' now a keyword

Syntax Error: Lexer: Bad token "\"I'm an \"\"infix\"\" symbol\""
```

Author: I don't know, tried a few things but didn't manage to make it work. Maybe @herbelin would have an idea?
In any case, this should be transformed to an actual example with the coqtop syntax to avoid breaking it by accident.

Reviewer 4: Looks like strings with spaces are not allowed as keywords, so they only work when used in a position which doesn't force a keyword.
For instance https://github.com/coq/coq/blob/a8ef679aedb5860a9817e0b8172d9f9cb69d0fda/test-suite/output/NotationSyntax.v#L29-L30 (keyword string with no space) works, but breaks if we add a space,
and https://github.com/coq/coq/blob/a8ef679aedb5860a9817e0b8172d9f9cb69d0fda/test-suite/output/NotationSyntax.v#L24-L25 (non keyword) works even if we add a space

This is probably a bug.

.. coqtop:: in

   Reserved Notation "A /\ B" (at level 80, right associativity).
   Reserved Notation "A \/ B" (at level 85, right associativity).

By default, a notation is considered nonassociative, but the
precedence level is mandatory (except for special cases whose level is
canonical, c.f. :ref:`next section <NotationFactorization>`).
The level is either a number or the phrase ``next level``
whose meaning is obvious.
Some :ref:`associativities are predefined <init-notations>` in the
``Notations`` module.

> .. TODO I don't find it obvious -- CPC

Author: I tried to add a simple sentence (it's indeed a bit confusing since `next` actually means `-1`)

.. coqtop:: in

   Reserved Notation "A /\ B" (at level 80, right associativity).
   Reserved Notation "A \/ B" (at level 85, right associativity).

By default, a notation is considered nonassociative, but the
precedence level is mandatory (except for special cases whose level is
canonical, c.f. :ref:`next section <NotationFactorization>`).
The level is either a number or the phrase ``next level``
whose meaning is obvious.
Some :ref:`associativities are predefined <init-notations>` in the
``Notations`` module.

.. TODO I don't find it obvious -- CPC

Sometimes, levels have to be voluntarily lowered, for instance if one tries

.. coqtop:: in

   Fail Reserved Notation "{ x : A | P }".

there is a conflict with the notation for
type casts. The notation for type casts, as shown by the command :cmd:`Print
Grammar` `constr` is at level 100. To avoid ``x : A`` being parsed as a type cast,
it is necessary to put ``x`` at a level below 100, typically 99. Hence, a correct
definition is the following:

.. coqtop:: reset all

>    Reserved Notation "{ x : A | P }" (x at level 99).

Author: Levels are associated to only tow things:
* whole rules with `at level`
* parameters (i.e., non terminal symbols) with `<parameter> at level`, only rule of level less or equal that parameter level will be used to parse that parameter (well, baring use of the recovery mechanism at least)

Putting levels to terminal symbols wouldn't have any meaning. I guess, this should be made clearer. Feel free to suggest where / a wording.

   Fail Reserved Notation "x << y << z" (at level 70, y at level 200).

In order to factorize the left part of the rules, the subexpression
referred to by ``y`` has to be at the same level in both rules. However the
default behavior puts ``y`` at the next level below 70 in the first rule
(``no associativity`` is the default). To fix this, we
need to force the parsing level of ``y``, as follows.

.. coqtop:: reset all

   Reserved Notation "x << y" (at level 70).
   Reserved Notation "x << y << z" (at level 70, y at next level).

Or better yet, simply let the defaults ensure the best factorization.

.. coqtop:: reset all

   Reserved Notation "x << y" (at level 70).
   Reserved Notation "x << y << z".
   Print Notation "_ << _ << _".

For the sake of factorization with Coq predefined rules, simple rules
have to be observed for notations starting with a symbol, e.g., rules
starting with “\ ``{``\ ” or “\ ``(``\ ” should be put at level 0. The list
of Coq predefined notations can be found in the chapter on :ref:`thecoqlibrary`.

.. warn:: Closed notations (i.e. starting and ending with a terminal symbol) should usually be at level 0 (default).
   :name: closed-notation-not-level-0

>    It is usually better to put closed notations, that is the ones starting and ending with a terminal symbol, at level 0.

Reviewer 3: Why is that so? What does this "usually" hides?

Author: Here usually means almost always, except in a few very specific cases that I don't precisely remember.

   The default precedence level for the inner sub-expressions of closed notations is 200, and the default
   level for the notation itself is 0. E.g.,

   .. coqtop:: in

      Reserved Notation "( x , y )".

   is essentially equivalent to

   .. coqtop:: reset in warn

      Reserved Notation "( x , y )" (at level 0, x, y at level 200).

.. warn:: Postfix notations (i.e. starting with a nonterminal symbol and ending with a terminal symbol) should usually be at level 1 (default).")
   :name: postfix-notation-not-level-1

>    It is usually better to put postfix notations, that is the ones ending with a terminal symbol, at level 1.

Reviewer 3: Same comment as above.

Author: same meaning

   scope (see :ref:`Scopes`), either the scope has to be opened or a
   delimiter has to exist in the scope for the notation to be usable.

Custom entries

> ~~~~~~~~~~~~~~

Author: Indeed, tried to improve that.
